content/post/2018-07-01-Automate-your-grammar-checks.md

```diff
@@ -0,0 +1,92 @@
+OK, to do this I came up with script below. It iterates over every file with `.md` extension and runs Hunspell on that file. Next it sorts output, removes duplicate lines and ones starting with `*` or `@(#)`. After that it parses Hunspell output to create human readable output with some colors. Unfortunately I didn't found how to get Hunspell to return different exit code than 0, so the script looks for phrase "Misspelled word" in output and returns 1 if such phrase was detected.
+
+```bash
+#!/bin/bash
+
+RET=0
+
+exec 5>&1
+for file in $(find . -name "*.md"); do
+    echo -e "\n\e[33mChecking $file\e[0m"
+    OUT=$(cat $file | \
+        hunspell -a -d en_US 2>/dev/null | \
+        sort --ignore-case | uniq | \
+        grep -v '^\*$\|@(#).*' | \
+        sed -r "s/^[&#]\s([[:alpha:]]+)\s([0-9]+)\s([0-9]+):/$(printf '\033[0m')Misspelled word $(printf '\033[31m\033[1m')\1$(printf '\033[0m') in line \2 position \3, possible alternatives:$(printf '\033[36m')/g" | \
+        tee >(cat - >&5))
+    echo -en "\e[0m"
+    if [ "$(echo $OUT | grep 'Misspelled word' | wc -l)" -gt 0 ]; then
+        RET=1
+    fi
+done
+
+exit $RET
+```
+
+### Not so markdown friendly Hunspell
+
+After creating above script and running it on some markdown files with embedded code (like this blog post) it was clear that I need some other tool. Script produced many more lines which were part of code section and weren't important, for example:
+```
+& ' 15 29: e, s, i, a, n, r, t, o, l, c, d, u, g, m, p
+& ' 15 70: e, s, i, a, n, r, t, o, l, c, d, u, g, m, p
+& 'Misspelled 3 29: misspelled, dispelled, misspell
+```
+It was time to get back to the drawing board.
+
+## Developers lingo
+
+Maybe I started it wrong? I essentially need a tool written by developers for developers not to "spell check" my blog posts, but to "lint" my "code". And since my "code" looks more like a GitHub read-me or a documentation, I should probably look for a "documentation linter". Quick google search and here I found a Codeship post about [improving documentation by automating spelling and grammar checks](https://blog.codeship.com/improve-documentation-by-automating-spelling-and-grammar-checks/). Hell yeah! That is exactly what I want to do! That blog post describes how to use an awesome tool called ["markdown-spellcheck"](https://www.npmjs.com/package/markdown-spellcheck) which was created to run spell checks on markdown files (duh!) using Hunspell and it even supports non-interactive mode! After running simple `npm install markdown-spellcheck -g` and `mdspell --ignore-acronyms --en-us --report "content/*/*.md"` I've got a nicely formatted output:
+
+![mdspell example output](/images/20180701-mdspell-example.png)
+
+Since not every word can be in dictionary, so `mdspell` can use `.spelling` file to store exceptions. Write a word in this file and `mdspell` won't report it as an error.
```

Comment:
Since not every word can be in dictionary, `mdspell` can use `.spelling` file to store exceptions.